modules/ROOT/edits/execution-engine.adoc

```diff
@@ -0,0 +1,152 @@
+misusing the CPU Light thread pool. This issue can be quickly checked by taking a
+thread dump of the runtime and looking for `WAITING` or `BLOCKED` or for
+long-running processes in the CPU Light threads.
+
+=== CPU Intensive
+
+CPU Intensive is also a small thread pool (2 threads per available core by default),
+but it provides a queue for accepting more tasks.
+
+=== IO
+
+IO is an elastic thread pool that grows as needed.
+
+Tasks running in this pool should spend most of their time `WAITING` or `BLOCKED`
+instead of doing CPU work, so they do not compete with the other pools.
+
+Also, when a transaction is active (since many transaction managers require all
+steps of the same transaction to be handled by the same thread), the `IO` pool
+is used.
+
+=== Custom Pools
+
+Apart from the three core thread pools, Mule or some components might create additional pools for specific purposes, such as:
+
+* NIO Selectors +
+Enables non-blocking IO. Each connector can create as many as required.
+* Recurring tasks pools +
+Some connectors or components might create specific pools to perform recurring tasks (e.g: expiration monitors, queue consumers, etc.)
+
+=== Proactor Pattern
+
+Proactor is a design pattern for asynchronous execution. To understand how the Proactor design pattern works, visit https://en.wikipedia.org/wiki/Proactor_pattern
+
+According to this design pattern, all tasks are classified in categories that correspond to each of Mule thread pools, and each task is submitted for execution to its corresponding thread pool.
+
+Consider the following example where a flow pulls down a JSON array of Person objects described in JSON format, pushes the content through an HTTP request, picks the name of the first entry and does some additional processing:
+
+[source, xml, linenums]
+----
+<flow>
+  <sftp:read path="personsArray.json" /> <1>
+  <http:request path="/persons" method="POST" /> <2>
+  <set-variable variableName="firstEntryName" value="#[payload[0].name]" /> <3>
+  <ee:transform ... /> <4>
+  <logger message="#[vars.firstEntryName]" /> <5>
+</flow>
+----
+
+According to the Proactor pattern, Mule submits the tasks as follows:
+
+<1> The blocking operation(`<sftp:read>`) executes in the IO pool.
+<2> `<http:request>` is a non blocking operation. Mule performs the request in the current thread. When the flow receives the response, Mule switches to the `CPU_LIGHT` pool.
+<3> `<set-variable>` operations should be fairly quick and stay in `CPU_LIGHT`. There is no thread switch.
+<4> `<ee:transform>`, this is potentially a computational heavy transformation, so Mule switches to the `CPU_INTENSIVE` pool.
+<5> Logger stays on `CPU_INTENSIVE`. There is no thread switch. +
+[NOTE]
+Due to optimizations regarding latency, thread switches are omitted when an IO or `CPU_INTENSIVE` task is followed by a `CPU_LIGHT` one. The reasoning behind this optimization is that executing said `CPU_LIGHT` task is most likely cheaper than the thread switch.
```

Comment:
```suggestion
Due to latency optimizations, thread switches are omitted when an I/O or `CPU_INTENSIVE` task is followed by a `CPU_LIGHT` task, because executing the `CPU_LIGHT` task is most likely cheaper than performing a thread switch.
```